Add optional support for closures in cached config

bcremer opened the pull request:
This adds support for exporting closures into the cache config file
using `brick/varexporter`.

Fixes #6.


|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | yes
| QA            | yes

### Description

Fixes #6.

bcremer committed:
Add optional support for closures in cached config

This adds support for exporting closures into the cache config file
using brick/varexporter as an optional (soft) dependency.

Fixes #6.

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer committed:
Include optional brick/varexporter dependency in coverage travis run

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer commented:
The optional `brick/varexporter` dependency is currently only tested for one run in the test matrix.
Having the package as hard dependency would probably mitigate bugs in weird combinations of the PHP Version and `brick/varexporter`. 

We could contraint the lower bound of `brick/varexporter` with a composer conflict declaration.

bcremer commented on test/ConfigAggregatorTest.php:
```diff
@@ -23,7 +24,6 @@
 class ConfigAggregatorTest extends TestCase
 {
     private $cacheFile;
-    private $lockFile;
```
Unused private property.

bcremer commented on test/ConfigAggregatorTest.php:
```diff
@@ -41,13 +41,13 @@ protected function tearDown()
     }
 
-    public function testConfigAggregatorRisesExceptionIfProviderClassDoesNotExist()
+    public function testConfigAggregatorRaisesExceptionIfProviderClassDoesNotExist()
```
Unrelated Change: Typo Rises -> Raises

bcremer commented on test/ConfigAggregatorTest.php:
```diff
@@ -92,10 +92,61 @@ function () {
+            );
+        }
+
+        $this->expectException(ConfigCannotBeCachedException::class);
```
This is actually an upside of using `Brick\VarExporter`.
The exception is thrown early during cache generation and not during the second request when the (unusable) cache file is read.

bcremer commented on src/ConfigCannotBeCachedException.php:
```diff
@@ -0,0 +1,34 @@
+        return new self(
+            sprintf(
+                'Cannot export config into a cache file. Config contains uncacheable entries: %s',
+                $exportException->getMessage()
```
Should the underlying exception message leak into the message?

bcremer commented on src/ConfigCannotBeCachedException.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        return new self(
+            sprintf(
+                'Cannot export config into a cache file. Config contains uncacheable entries: %s',
```
Should the underlying exception message leak into the message?

weierophinney replied:
Does it contain information about the uncacheable entries? And is there any other way to get at that information?

If the answer to the first question is yes and the second is no, then this is the right approach, from a "can I diagnose what is wrong from the exception message?" perspective.

bcremer replied:
The underlying exception message contains useful debug information, so I'll leave it in the message.

bcremer commented on src/ConfigAggregator.php:
```diff
@@ -274,12 +277,18 @@ private function cacheConfig(array $config, $cachedConfigFile)
-            date('c'),
-            var_export($config, true)
-        );
+        $hasVarExporter = is_callable("Brick\VarExporter\VarExporter::export");
```
There might be room for improvement here by using class_exists, but this code is not performance relevant so anything that does work with PHP 5.6 and does not produce too many side effects should do the trick.

bcremer commented:
As per https://github.com/laminas/technical-steering-committee/blob/master/meetings/minutes/2020-03-02-TSC-Minutes.md#summary-2 would it be appropriate to drop all PHP versions less than 7.3 from the build matrix?

weierophinney commented:
> would it be appropriate to drop all PHP versions less than 7.3 from the build matrix?

The questions to answer are: Will you be leveraging features from PHP 7.3, and/or will some features you use from other libraries not work on prior versions?

Those are the typical reasons to bump the PHP version. If the answer to either of those is "yes", let's bump.

weierophinney commented:
> Having the package as hard dependency would probably mitigate bugs in weird combinations of the PHP Version and `brick/varexporter`.

You mentioned that you thought nikic/php-parser was a "heavy" dependency. I'm not quite sure what you mean by that, but I generally think of a dependency as "heavy" if it in turn pulls in many other packages as dependencies, and/or has stricter requirements (PHP version, extensions, etc.) than I might otherwise require. In this particular case, I disagree with that assessment: the package depends on zero other packages, only requires PHP 7.0 or greater, and the tokenizer extension (which is in vanilla builds of PHP, and present in the "common" metapackage of most distro packages). brick/varexporter is similar: the only package it depends on is nikic/php-parser, and it bumps the minimum PHP version to 7.2 (at least in current HEAD).

As such, I see no reason NOT to make it a package requirement, especially considering all we gain. This **will** require us bumping to PHP 7.2 at the minimum, due to the requirements of brick/varexporter; per our release policy, that's acceptable for new minor releases, which this already targets.

weierophinney commented on src/ConfigAggregator.php:
```diff
@@ -274,12 +277,18 @@ private function cacheConfig(array $config, $cachedConfigFile)
+                self::CACHE_TEMPLATE,
+                get_class($this),
+                date('c'),
+                $hasVarExporter ? VarExporter::export($config) : var_export($config, true)
```
I would argue we should pass the following flags to `VarExporter::export()`:

- `VarExporter::ADD_RETURN` would allow us to simplify the template
- `VarExporter::CLOSURE_SNAPSHOT_USES` would allow usage of `use` with closures and ensure arrow functions that are using variables from the current context work without issue. (I know I've actually done this a few times!)

bcremer replied:
Good suggestions, added both flags.

weierophinney commented on test/ConfigAggregatorTest.php:
```diff
@@ -92,10 +92,61 @@ function () {
         $this->assertEquals(['foo' => 'bar', ConfigAggregator::ENABLE_CACHE => true], $cachedConfig);
     }
 
+    public function testConfigAggregatorCanCacheConfigWithClosures()
```
Assuming we add the `CLOSURE_SNAPSHOT_USES` option, we should have another test to validate that a closure that uses binding is serialized correctly.

bcremer replied:
Added testcase.

weierophinney reviewed (changes requested):
I'm really excited for this!

As noted in other comments:

- Let's make brick/varexporter a required dependency...
- Which means you can get rid of some conditionals and test skipping clauses.
- Add some config options when exporting, so we can use bound closures, and...
- simplify the cache template, as we can have brick/varexporter create a `return`able config.

bcremer committed:
Bumb minimum required PHP Version to 7.2

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer committed:
Require brick/varexporter as dependency, drop from suggest

As brick/varexporter is now always available some conditional checks in
the library code and the testcases can be dropped.

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer committed:
Simplify the template by using VarExporter::ADD_RETURN

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer committed:
Use VarExporter::CLOSURE_SNAPSHOT_USES flag, add testcase

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer committed:
Update phpunit to a supported version

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

bcremer commented:
@weierophinney Thanks for your review.
I have implemented all your suggestions as separate commits.

samsonasik commented on test/ConfigAggregatorTest.php:
```diff
@@ -90,10 +90,69 @@ function () {
+        $cachedConfig = include $this->cacheFile;
+
+        $this->assertIsCallable($cachedConfig['toUpper']);
+        $this->assertSame('FOOBAR', call_user_func($cachedConfig['toUpper'], 'foobar'));
```
`call_user_func()` usage should can be changed with direct call: 

```suggestion
        $this->assertSame('FOOBAR', $cachedConfig['toUpper']('foobar'));
```

ref https://3v4l.org/49DtW

bcremer replied:
Fixed in my latest commit.

samsonasik commented on test/ConfigAggregatorTest.php:
```diff
@@ -90,10 +90,69 @@ function () {
+        $cachedConfig = include $this->cacheFile;
+
+        $this->assertIsCallable($cachedConfig['addPrefix']);
+        $this->assertSame('prefixfoobar', call_user_func($cachedConfig['addPrefix'], 'foobar'));
```
`call_user_func()` usage should can be changed with direct call: 

```suggestion
        $this->assertSame('prefixfoobar', $cachedConfig['addPrefix']('foobar'));
```

ref https://3v4l.org/IGiDm

bcremer committed:
Use some recent language fatures for private methods.

Signed-off-by: Benjamin Cremer <benjamin.cremer@check24.de>

weierophinney committed:
qa: add visibility modifiers to existing constants

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: indentation in new tests

Ensure arrays are indented correctly, and have trailing commas when
spanning multiple lines.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: adds CHANGELOG entry for #7

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney reviewed (approved):
I rebased against current develop, and did a little CS/QA housekeeping (including adding changelog entries). Looks great - I'll be releasing shortly!
